database_test.go

```diff
@@ -469,15 +497,48 @@ func TestDatabaseUpdate(t *testing.T) {
 				"a4", "5d6506974928a003d2a0ccbd7a40b5341ad10578fd3f54527087c5ecbbd17a12"),
 		},
 	}
-	db.Update(mockAPI)
-	if db.err == nil {
+	updated, delay = db.Update(mockAPI)
+	if db.err == nil || updated {
 		t.Fatalf("update 5, unexpected update success")
 	}
+	if math.Abs((config.UpdatePeriod - delay).Seconds()) > 31 {
+		t.Fatalf("update 5, delay jitter was more than 30 seconds")
+	}
 	gotDB = &database{last: db.last, tfu: db.tfu, tfl: db.tfl}
 	wantDB = &database{}
 	if !reflect.DeepEqual(gotDB, wantDB) {
 		t.Fatalf("update 5, database state mismatch:\ngot  %+v\nwant %+v", gotDB, wantDB)
 	}
+
+	// Update 6: api is broken for some unknown reason. Checks the backoff
+	errResponse = errors.New("Something broke")
+	updated, delay = db.Update(mockAPI)
+	if db.err == nil || updated {
+		t.Fatalf("update 6, unexpected update success")
+	}
+	minDelay := baseRetryDelay.Seconds() * float64(1) * float64(db.updateAPIErrors)
+	maxDelay := baseRetryDelay.Seconds() * float64(2) * float64(db.updateAPIErrors)
+	if delay.Seconds() < minDelay || delay.Seconds() > maxDelay {
+		t.Fatalf("update 6, Expected delay %v to be between %v and %v", delay.Seconds(), minDelay, maxDelay)
+	}
+	updated, delay = db.Update(mockAPI)
+	if db.err == nil || updated {
+		t.Fatalf("update 6, unexpected update success")
+	}
+	minDelay = baseRetryDelay.Seconds() * float64(1) * float64(db.updateAPIErrors)
+	maxDelay = baseRetryDelay.Seconds() * float64(2) * float64(db.updateAPIErrors)
+	if delay.Seconds() < minDelay || delay.Seconds() > maxDelay {
+		t.Fatalf("update 6, Expected delay %v to be between %v and %v", delay.Seconds(), minDelay, maxDelay)
+	}
+	updated, delay = db.Update(mockAPI)
+	if db.err == nil || updated {
+		t.Fatalf("update 6, unexpected update success")
+	}
+	minDelay = baseRetryDelay.Seconds() * float64(1) * float64(db.updateAPIErrors)
```

Comment:
This is not quite the same formula that is used to actually calculate the delay, but it just so happens to work (most of the time). In this third consecutive failure, minDelay is `3*baseRetryDelay` and maxDelay is `6*baseRetryDelay`. Since the multiplier is supposed to be `2^(n-1) * (1+rand())`, it should be more like `4*baseRetryDelay` to `8*baseRetryDelay`.

For these particular tests, I think it would be useful to just figure out these constants (i.e. `minDelay = baseRetryDelay.Seconds() * 4`) instead of re-calculating from internal variables. These tests would also pass if db.updateAPIErrors was never incremented.